### PR TITLE
Record.store(), Record.import(), and roles API support

### PR DESCRIPTION
**Pull request by csuwildcat**
First pass at PR

### CODE REVIEW

**Comment by frankhinek** [comment]
@csuwildcat Node and one of the browser tests are failing.  Can you look into why?

File: packages/api/src/record.ts [hunk 1]
```diff
@@ -66,6 +69,20 @@ export type RecordUpdateOptions = {
...
+    const val = obj[key];
+    if (val === undefined) {
+      delete obj[key];
+    } else if (typeof val === 'object') {
+      removeUndefinedProperties(val);
+    }
+  });
+  return obj;
+}
+
```
Comment (frankhinek) [inline packages/api/src/record.ts:85]: To avoid duplication use the function that [exists](https://github.com/TBD54566975/web5-js/blob/2bd93a85754b6f39208968ce2fc464bf943f15c2/packages/common/src/object.ts#L35) in `@web5/common`.

`import { removeUndefinedProperties } from '@web5/common';`

File: packages/api/tests/record.spec.ts [hunk 2]
```diff
@@ -1345,7 +1465,7 @@ describe('Record', () => {
     });
 
     // TODO: Fix after changes are made to dwn-sdk-js to include the initial write in every query/read response.
```
Comment (frankhinek) [inline packages/api/tests/record.spec.ts:1467]: TODO should be removed now that this test is passing.
```suggestion
```

File: packages/api/tests/record.spec.ts [hunk 3]
```diff
@@ -97,6 +97,127 @@ describe('Record', () => {
...
+    //   from    : bobDid.did,
+    //   message : {
+    //     filter: {
+    //       schema: 'http://email-protocol.xyz/schema/email'
+    //     }
+    //   }
+    // });
+    // expect(bobQueryResult.status.code).to.equal(200);
+    // expect(bobQueryResult.records).to.exist;
+    // expect(bobQueryResult.records!.length).to.equal(1);
```
Comment (frankhinek) [inline packages/api/tests/record.spec.ts:218]: Why is this included but commented out?

File: packages/agent/src/dwn-manager.ts [hunk 4]
```diff
@@ -242,17 +242,18 @@ export class DwnManager {
   }
 
   private async constructDwnMessage(options: {
-    request: ProcessDwnRequest
+    request : ProcessDwnRequest
```
Comment (frankhinek) [inline packages/agent/src/dwn-manager.ts:245]: Unnecessary space added in this PR that should be removed.
```suggestion
    request: ProcessDwnRequest
```

File: packages/api/src/record.ts [hunk 5]
```diff
@@ -77,16 +94,45 @@ export class Record implements RecordModel {
...
+  private _initialWrite: RecordOptions['initialWrite'];
+  private _initialWriteStored: boolean;
   private _recordId: string;
-
+  private _protocolRole: RecordOptions['protocolRole'];
   // Getters for immutable DWN Record properties.
 
+  static sendCache = new Map();
+  static sendCacheLimit = 100;
+  static setSendCache(recordId, target){
```
Comment (frankhinek) [inline packages/api/src/record.ts:109]: Please add some documentation that explains what this does.
```suggestion
  static setSendCache(recordId, target) {
```

File: packages/api/src/record.ts [hunk 6]
```diff
@@ -295,6 +350,91 @@ export class Record implements RecordModel {
...
+
+    if (options.dataStream) {
+      request.dataStream = options.dataStream;
+    }
+
+    return this._agent.processDwnRequest(request);
+  }
+
+
+  async store(options: any = {}): Promise<any> {
```
Comment (frankhinek) [inline packages/api/src/record.ts:381]: Please add some documentation that explains what this does.

```suggestion
  async store(options: any = {}): Promise<any> {
```

File: packages/api/src/record.ts [hunk 7]
```diff
@@ -295,6 +350,91 @@ export class Record implements RecordModel {
     return dataObj;
   }
 
+  private async _processMessage(options: any){
+
+    const { store = true } = options;
```
Comment (frankhinek) [inline packages/api/src/record.ts:355]: Please add some documentation that explains why another `processMessage()` private method was added.

File: packages/api/src/record.ts [hunk 8]
```diff
@@ -304,14 +444,55 @@ export class Record implements RecordModel {
...
-      target         : target,
-      messageOptions : this.toJSON(),
-    });
+  async send(_options?: string | { [key: string]: any }): Promise<ResponseStatus> {
+
+    const initialWrite = this._initialWrite;
+    const options = !_options ? { target: this._connectedDid } : typeof _options === 'string' ? { target: _options } : _options;
+
+    if (!options.target){
+      options.target = this._connectedDid;
```
Comment (frankhinek) [inline packages/api/src/record.ts:453]: This is some ugly looking code 😄 

It seems like there should be a better way to accomplish this.

File: packages/api/src/record.ts [hunk 9]
```diff
@@ -304,14 +444,55 @@ export class Record implements RecordModel {
    *
    * @beta
    */
```
Comment (frankhinek) [inline packages/api/src/record.ts:446]: Please update the explanation for what `send()` does because it isn't the same as before.

File: packages/api/src/record.ts [hunk 10]
```diff
@@ -304,14 +444,55 @@ export class Record implements RecordModel {
...
+        recordId      : this._recordId,
+        descriptor    : this._descriptor,
+        attestation   : this._attestation,
+        authorization : this._authorization,
+        encryption    : this._encryption,
+      });
+    }
+    else {
+      latestState.messageOptions = this.toJSON();
+    }
```
Comment (frankhinek) [inline packages/api/src/record.ts:493]: Please add inline comments that explain what is going on here so its easier to maintain in the future.

File: packages/api/src/record.ts (hunk 1 above)
Comment (csuwildcat) [reply packages/api/src/record.ts:93]: Many tests fail when I switch to using the web5/common packages version.

Comment (csuwildcat) [reply packages/api/src/record.ts:93]: I am guessing it is because it assumes it will be a typed object of the one specified in the arguments of that version?

File: packages/api/src/record.ts (hunk 5 above)
Comment (csuwildcat) [reply packages/api/src/record.ts]: Added

File: packages/api/src/record.ts (hunk 8 above)
Comment (csuwildcat) [reply packages/api/src/record.ts]: Liran changed it

File: packages/api/tests/record.spec.ts (hunk 3 above)
Comment (csuwildcat) [reply packages/api/tests/record.spec.ts]: Removed old unused test.

File: packages/agent/src/dwn-manager.ts (hunk 4 above)
Comment (csuwildcat) [reply packages/agent/src/dwn-manager.ts]: Fixed

File: packages/api/src/record.ts (hunk 7 above)
Comment (csuwildcat) [reply packages/api/src/record.ts]: Done

File: packages/api/src/record.ts (hunk 10 above)
Comment (csuwildcat) [reply packages/api/src/record.ts]: Done

File: packages/api/src/record.ts (hunk 6 above)
Comment (csuwildcat) [reply packages/api/src/record.ts]: Added

**Comment by LiranCohen** [comment]
@frankhinek Would love another run through and blessing from you.

There is more clean up to do, specifically I'd love to get rid of `toJSON()` completely in favor of the `rawMessage` getter.
I'd like to simplify the messageOptions as well for record creation and re-think some of this flow now that we have more context.

But I think this is in a good place to un-block some bugs. And do further cleanup in subsequent PRs.

Would also love feedback from the community and users if the `import()` and `store()` methods make sense from a dev-ex perspective.

**Review by frankhinek — APPROVED** [review]

**Comment by frankhinek** [comment]
> Would also love feedback from the community and users if the `import()` and `store()` methods make sense from a dev-ex perspective.

This is the biggest open question I still have.  Something about `import()` still seems confusing but better to ship and get feedback than endless deliberate.  We can always change it.

**Review by LiranCohen — APPROVED** [review]
